README.md
# Node Token Authentication

This repo uses JSON Web Tokens and the [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken) package to implement token based authentication on a simple Node.js API.

This is a starting point to demonstrate the method of authentication by verifying a token using Express route middleware.

## Requirements

- node and npm

## Usage

1. Clone the repo: `git clone git@github.com:abr4xas/node-token-authentication`
2. Install dependencies: `npm install`
3. Change SECRET in `config.js`
4. Add your own MongoDB database to `config.js`
5. Start the server: `node server.js`
6. Create sample user by visiting: `http://localhost:8080/setup`

Once everything is set up, we can begin to use our app by creating and verifying tokens.

### Getting a Token

Send a `POST` request to `http://localhost:8080/api/authenticate` with test user parameters as `x-www-form-urlencoded`.

```
  {
    name: 'abr4xas',
    password: '123456'
  }
```

### Verifying a Token and Listing Users

Send a `GET` request to `http://localhost:8080/api/users` with a header parameter of `x-access-token` and the token.

You can also send the token as a URL parameter: `http://localhost:8080/api/users?token=YOUR_TOKEN_HERE`

Or you can send the token as a POST parameter of `token`.


CHANGELOG
* [31mcbfdec8[m - [32mabr4xas[m - [33m (HEAD -> master, origin/master, origin/HEAD)[m actualizando aplicacion :bug: [32m(8 hours ago)[m
* [31mac600e2[m - [32mabr4xas[m - [33m[m actualizando repo [32m(8 hours ago)[m
* [31m1f6412e[m - [32mabr4xas[m - [33m[m fix jsonwebtoken: expiresInMinutes and expiresInSeconds is deprecated. [32m(8 hours ago)[m
* [31ma7b60e8[m - [32mabr4xas[m - [33m[m actualizando aplicacion :bug: [32m(9 hours ago)[m
*   [31m8cfdc8a[m - [32mChris Sevilleja[m - [33m[m Merge pull request #1 from shahar1/patch-1 [32m(6 months ago)[m
|\  
| * [31mfd5f734[m - [32mshahar1[m - [33m[m Update server.js [32m(6 months ago)[m
|/  
* [31me90ea83[m - [32mChris Sevilleja[m - [33m[m Update server.js [32m(6 months ago)[m
* [31m335b8a2[m - [32mChris Sevilleja[m - [33m[m Update README.md [32m(11 months ago)[m
* [31mc2fb1fd[m - [32mChris Sevilleja[m - [33m[m fixing token grab [32m(11 months ago)[m
* [31m2840906[m - [32mChris Sevilleja[m - [33m[m cleaning [32m(11 months ago)[m
* [31mdf3dc97[m - [32mChris Sevilleja[m - [33m[m spacing [32m(11 months ago)[m
* [31mfea3586[m - [32mChris Sevilleja[m - [33m[m final [32m(11 months ago)[m
* [31m4ec1aad[m - [32mChris Sevilleja[m - [33m[m available in check [32m(11 months ago)[m
* [31me4ec1f1[m - [32mChris Sevilleja[m - [33m[m more work [32m(11 months ago)[m
* [31m58e04cc[m - [32mChris Sevilleja[m - [33m[m whyyy [32m(11 months ago)[m
* [31m899e230[m - [32mChris Sevilleja[m - [33m[m doing things [32m(12 months ago)[m
* [31mf1edf4e[m - [32mChris Sevilleja[m - [33m[m setting up [32m(12 months ago)[m
* [31m2d843c0[m - [32mChris Sevilleja[m - [33m[m Initial commit [32m(12 months ago)[m

server.js
// =================================================================
// get the packages we need ========================================
// =================================================================
var express 	= require('express'),
    app         = express(),
    bodyParser  = require('body-parser'),
    morgan      = require('morgan'),
    mongoose    = require('mongoose'),
    jwt    = require('jsonwebtoken'),
    config = require('./config'),
    User   = require('./app/models/user'); // get our mongoose model

// =================================================================
// configuration ===================================================
// =================================================================
var port = process.env.PORT || 8080; // used to create, sign, and verify tokens
mongoose.connect(config.database); // connect to database
app.set('superSecret', config.secret); // secret variable

// use body parser so we can get info from POST and/or URL parameters
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());

// use morgan to log requests to the console
app.use(morgan('dev'));

// =================================================================
// routes ==========================================================
// =================================================================
app.get('/setup', function(req, res) {

    // create a sample user
    var nick = new User({
        name: 'abr4xas',
        password: '123456',
        admin: true
    });
    nick.save(function(err) {
        if (err) throw err;

        console.log('User saved successfully');
        res.json({ success: true });
    });
});

// basic route (http://localhost:8080)
app.get('/', function(req, res) {
    res.send('Hello! The API is at http://localhost:' + port + '/api');
});

// ---------------------------------------------------------
// get an instance of the router for api routes
// ---------------------------------------------------------
var apiRoutes = express.Router();

// ---------------------------------------------------------
// authentication (no middleware necessary since this isnt authenticated)
// ---------------------------------------------------------
// http://localhost:8080/api/authenticate
apiRoutes.post('/authenticate', function(req, res) {

    // find the user
    User.findOne({
        name: req.body.name
    }, function(err, user) {

        if (err) throw err;

        if (!user) {
            res.json({ success: false, message: 'Authentication failed. User not found.' });
        } else if (user) {

            // check if password matches
            if (user.password != req.body.password) {
                res.json({ success: false, message: 'Authentication failed. Wrong password.' });
            } else {

                // if user is found and password is right
                // create a token
                var token = jwt.sign(user, app.get('superSecret'), {
                    expiresIn: 1440 // expires in 24 hours
                });

                res.json({
                    success: true,
                    message: 'Enjoy your token!',
                    token: token
                });
            }

        }

    });
});

// ---------------------------------------------------------
// route middleware to authenticate and check token
// ---------------------------------------------------------
apiRoutes.use(function(req, res, next) {

    // check header or url parameters or post parameters for token
    var token = req.body.token || req.params('token') || req.headers['x-access-token'];

    // decode token
    if (token) {

        // verifies secret and checks exp
        jwt.verify(token, app.get('superSecret'), function(err, decoded) {
            if (err) {
                return res.json({ success: false, message: 'Failed to authenticate token.' });
            } else {
                // if everything is good, save to request for use in other routes
                req.decoded = decoded;
                next();
            }
        });

    } else {

        // if there is no token
        // return an error
        return res.status(403).send({
            success: false,
            message: 'No token provided.'
        });

    }

});

// ---------------------------------------------------------
// authenticated routes
// ---------------------------------------------------------
apiRoutes.get('/', function(req, res) {
    res.json({ message: 'Welcome to the coolest API on earth!' });
});

apiRoutes.get('/users', function(req, res) {
    User.find({}, function(err, users) {
        res.json(users);
    });
});

apiRoutes.get('/check', function(req, res) {
    res.json(req.decoded);
});

app.use('/api', apiRoutes);

// =================================================================
// start the server ================================================
// =================================================================
app.listen(port);
console.log('Magic happens at http://localhost:' + port);
